### DISCUSSION TITLE
Using parsed dependencies in SpanCategorizer suggester with transformers

### DISCUSSION THREAD

**Discussion by phlobo** [explosion/spaCy#10139] [Help: Coding & Implementations] [answered]
Hello!

I am implementing a custom suggester, that uses dependency parsing information (e.g., noun chunks) for suggesting spans to the SpanCategorizer.

I basically follow the config suggested here:
https://github.com/explosion/spaCy/discussions/10059

and added:

```
[nlp]
lang = "de"
pipeline = ["transformer", "parser", "spancat"]
```

```
[components.parser]
source = "de_dep_news_trf"
replace_listeners = ["model.tok2vec"]
```

as well as

```
[training]
accumulate_gradient = 3
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
patience = 1600
max_epochs = 100
max_steps = 20000
eval_frequency = 200
frozen_components = ["parser"]
annotating_components = ["parser"]
before_to_disk = null
```

In my custom suggester, I still get an error:
`ValueError: [E029] noun_chunks requires the dependency parse, which requires a statistical model to be installed and loaded.`

It has been suggested here:
https://github.com/explosion/spaCy/discussions/9201

that you need to add `attrs` to the `embed` section of `tok2vec`, but that section doesn't seem to exist for 

```
[components.spancat.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0
pooling = {"@layers":"reduce_mean.v1"}
upstream = "transformer"
```

**Comment by adrianeboyd** [comment, marked as answer]
Noun chunks require both `token.pos` and a parse, and POS usually comes from either `tagger+attribute_ruler` or `morphologizer` in the provided trained pipelines. (Sorry, the error message looks like it's gotten a bit out-of-date.)

For German, it's: `tok2vec/transformer`, `morphologizer`, `parser`. Both of those listen to the same transformer and you don't want to duplicate the `transformer` component that many times with `replace_listeners` (it would be both huge and slow), so instead it would be better to use a custom name for the spancat's `transformer` component and use that as `upstream` for the spancat listener instead. Put all the new components after the existing frozen components. So:

```ini
pipeline = ["transformer", "morphologizer", "parser", "transformer_spancat", "spancat"]

...

[components.spancat.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0
pooling = {"@layers":"reduce_mean.v1"}
upstream = "transformer_spancat"
```

You don't need to add `DEP` anywhere with the transformer, or to use noun chunks with spancat.

This is not anything we published officially and it's messy and inefficient, but I implemented a noun chunk suggester as a proof-of-concept here, which does noun chunk +/- two tokens as the suggested spans, look in `scripts/code.py`:

https://github.com/adrianeboyd/projects/tree/feature/ner-bootstrapped-confidence/experimental/ner_confidence

**Reply by phlobo** [reply]
Thank for your help! Unfortunately, I still cannot use the parsing results in my suggester.

This is my config:

```
[paths]
train = null
dev = null
vectors = null
init_tok2vec = null

[system]
gpu_allocator = "pytorch"
seed = 0

[nlp]
lang = "de"
pipeline = ["transformer", "morphologizer", "parser", "transformer_spancat", "spancat"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.morphologizer]
source = "de_dep_news_trf"

[components.parser]
source = "de_dep_news_trf"

[components.transformer]
source = "de_dep_news_trf"

[components.spancat]
factory = "spancat"
max_positive = 1
scorer = {"@scorers":"phlobo.flat_scorer"}
spans_key = "snomed"
threshold = 0.5

[components.spancat.model]
@architectures = "spacy.SpanCategorizer.v1"

[components.spancat.model.reducer]
@layers = "spacy.mean_max_reducer.v1"
hidden_size = 128

[components.spancat.model.scorer]
@layers = "spacy.LinearLogistic.v1"
nO = null
nI = null

[components.spancat.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0
pooling = {"@layers":"reduce_mean.v1"}
upstream = "transformer_spancat"

[components.spancat.suggester]
@misc = "phlobo.snomed_suggester"
sizes = [1, 2, 3, 4, 5]

[components.transformer_spancat]
factory = "transformer"
max_batch_items = 4096
set_extra_annotations = {"@annotation_setters":"spacy-transformers.null_annotation_setter.v1"}

[components.transformer_spancat.model]
@architectures = "spacy-transformers.TransformerModel.v3"
name = "deepset/gbert-base"
mixed_precision = false

[components.transformer_spancat.model.get_spans]
@span_getters = "spacy-transformers.strided_spans.v1"
window = 128
stride = 96

[components.transformer_spancat.model.grad_scaler_config]

[components.transformer_spancat.model.tokenizer_config]
use_fast = true

[components.transformer_spancat.model.transformer_config]


[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[training]
accumulate_gradient = 3
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
patience = 1600
max_epochs = 100
max_steps = 20000
eval_frequency = 200
frozen_components = ["transformer", "morphologizer", "parser"]
annotating_components = ["transformer", "morphologizer", "parser"]
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_padded.v1"
discard_oversize = true
size = 2000
buffer = 256
get_length = null

[training.logger]
@loggers = "spacy.WandbLogger.v3"
project_name = "ggponc_spacy"

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001

[training.optimizer.learn_rate]
@schedules = "warmup_linear.v1"
warmup_steps = 250
total_steps = 20000
initial_rate = 0.00005

[training.score_weights]
spans_sc_f = null
spans_sc_p = null
spans_sc_r = null
spans_snomed_f = 1.0
spans_snomed_p = 0.0
spans_snomed_r = 0.0
Diagnosis_or_Pathology_f = 0.0
Other_Finding_f = 0.0
Clinical_Drug_f = 0.0
Nutrient_or_Body_Substance_f = 0.0
External_Substance_f = 0.0
Therapeutic_f = 0.0
Diagnostic_f = 0.0


[pretraining]

[initialize]
vectors = ${paths.vectors}
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]
```

**Reply by adrianeboyd** [reply]
I don't see any obvious mistakes in the config. Can you provide more details about what's going wrong?

**Reply by phlobo** [reply]
It's the same error as before, I cannot call `noun_chunks` on the doc within the suggester:

`ValueError: [E029] noun_chunks requires the dependency parse, which requires a statistical model to be installed and loaded. For more info, see the documentation: ...`

This is what the suggester looks like:

```
@spacy.registry.misc("phlobo.snomed_suggester")
def build_chunk_and_ngram_suggester(sizes: List[int]) -> Suggester:

    def chunk_and_ngram_suggester(docs: Iterable[Doc], *, ops: Optional[Ops] = None) -> Ragged:
        if ops is None:
            ops = get_current_ops()
        spans = []
        lengths = []
        for doc in docs:
            for nc in doc.noun_chunks:
                ## do stuff
                pass
    
    return chunk_and_ngram_suggester
```

**Reply by phlobo** [reply]
Looking again at your suggester example, I added a check `if doc.has_annotation("DEP")` and some print statements. It looks like actually there are noun chunks assigned for some of the documents, but I do not understand why it is missing for others... In particular, since these others definitely have noun phrases that the `de_dep_news_trf` model is able to detect.

**Reply by phlobo** [reply]
After a bit more debugging, I noticed it's only a batch of 10 documents in the beginning that does not have the annotation! After these, all docs seem to have noun chunks. So I guess my problem is solved, but the behaviour is a bit puzzling

**Reply by adrianeboyd** [reply]
That sounds like it might be a bug, but it's kind of hard to tell without a complete example. I ran into a number of bugs related to empty docs or docs without suggestions while working on this, but I don't think that's directly related to the noun chunk error.

**Reply by phlobo** [reply]
I'm happy to share the project, but can only share the data privately - if that's an option, please let me know.

**Reply by adrianeboyd** [reply]
Ah, I think the problem is that the `initialize` step runs the suggester but the reference docs don't necessarily contain that annotation. I suspect there's no need to run the actual suggester and that running any simple suggester that suggests some spans would work just as well in that function.
